### PR TITLE
Return to Tiny: Reworks heights to use filters, allows spacemen to have varying heights

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

Re-pr of #66644 with some changes to get it working

image

- [x] ~~Huds get cut off~~ Resolved
- [x] ~~Very tall helmets get cut off~~ Resolved
- [x] Hair gets cut off
- [x] ~~Needs optimization before releasing to the world as a pref~~ jk lol

## Why It's Good For The Game

Look at that fella so cute

## Changelog

:cl: Melbert, That REALLY Good Soda Flavor, FatFat, AndreyGusev
add: Spacemen can now have varying height. (Admin only for now)
add: Dwarfs are now slightly shorter, but look way better.
/:cl:

### CODE REVIEW

**Comment by Kapu1178** [comment]
the filter abuse will almost certainly prove troublesome for low-end computers.

File: code/datums/datum.dm [hunk 1]
```diff
@@ -278,3 +282,72 @@
 /datum/proc/GenerateTag()
 	datum_flags |= DF_USE_TAG
 
+/datum/proc/add_filter(name,priority,list/params)
```
Comment (Mothblocks) [inline code/datums/datum.dm:285]: i dont understand, why are we doing this?

Comment (Mothblocks) [reply code/datums/datum.dm:285]: oh for images. ugh

**Comment by Farquaar** [comment]
Prelude to re-adding lavaland dwarves?

**Comment by necromanceranne** [comment]
He's just a little guy

**Comment by Pumpkinoe** [comment]
we are about to enter  new fucking era

**Comment by axietheaxolotl** [comment]
I can already tell that this can cause a cavalcade of issues with sprites. You can already tell the issues with the kneepads on the shortest one. Funny, but a pretty poor practical thing.
image

**Comment by Fikou** [comment]
> I can already tell that this can cause a cavalcade of issues with sprites. You can already tell the issues with the kneepads on the shortest one. Funny, but a pretty poor practical thing. image

dwarfs look shit already this is funny at least, and its not that bad

**Comment by SomeoneYouProbablyKnow** [comment]
Don't people already powergame dwarf mutation for the smaller sprite? Honestly that's just a consideration, nothing wrong with being able to tell sweaty gamers just from their sprite

**Comment by SomeoneYouProbablyKnow** [comment]
> > Don't people already powergame dwarf mutation for the smaller sprite? Honestly that's just a consideration, nothing wrong with being able to tell sweaty gamers just from their sprite
> 
> For the record the only one players will be able to obtain with this PR is Dwarfism, the other heights are admin only
> 
> But lol imagine powergaming a 2 pixel shorter height. Just call them "manlets"

5'11 vs 6'1

**Comment by Fikou** [comment]
can we make gigantism use this too?

**Comment by Pumpkinoe** [comment]
is there an idea of how long it may be until thisnis a preference?

**Comment by Mothblocks** [comment]
do not make this a preference

**Comment by Pumpkinoe** [comment]
please itd be so great though

**Comment by calliophis-maculiceps** [comment]
DO NOT MAKE THIS A PREFERENCE

**Comment by Pumpkinoe** [comment]
my dreams have been crushed utterly

**Comment by ChungusGamer666** [comment]
please make this a preference i need to roleplay my character being 5'11

File: code/datums/datum.dm (hunk 1 above)
Comment (ChungusGamer666) [reply code/datums/datum.dm:285]: this is extremely evil code and most likely not necessary at all to get this pr working

**Comment by Mothblocks** [comment]
its not going to happen bikeshed somewhere else

File: code/datums/datum.dm (hunk 1 above)
Comment (Mothblocks) [reply code/datums/datum.dm:285]: it's a side effect of oop only built on inheritance chains being terrible. most i could ask for on this is `TEST_ONLY_ASSERT(isatom(src) || istype(src, /image))` I think, which would at least highlight why this is being done beter

File: code/datums/datum.dm [hunk 2]
```diff
@@ -278,3 +282,101 @@
...
+	copied_parameters["priority"] = priority
+	filter_data[name] = copied_parameters
+	update_filters()
+
+/datum/proc/update_filters()
+	var/atom/atom_cast = src // this works with even images
+	atom_cast.filters = null
+	filter_data = sortTim(filter_data, GLOBAL_PROC_REF(cmp_filter_data_priority), TRUE)
+	for(var/f in filter_data)
+		var/list/data = filter_data[f]
```
Comment (jlsnow301) [inline code/datums/datum.dm:306]:
```suggestion
	for(var/filter in filter_data)
		var/list/data = filter_data[filter]
```

File: code/datums/datum.dm [hunk 3]
```diff
@@ -278,3 +282,101 @@
...
+ * * loop - loop arg of the BYOND animate() proc.
+ */
+/datum/proc/transition_filter(name, list/new_params, time, easing, loop)
+	var/filter = get_filter(name)
+	if(!filter)
+		return
+	animate(filter, new_params, time = time, easing = easing, loop = loop)
+	modify_filter(name, new_params)
+
+/datum/proc/change_filter_priority(name, new_priority)
```
Comment (jlsnow301) [inline code/datums/datum.dm:351]: please doc these (however trivial)

File: code/modules/mob/living/carbon/human/human_update_icons.dm [hunk 4]
```diff
@@ -705,6 +711,25 @@ generate/load female uniform sprites matching all previously decided variables
 	//eg: ammo counters, primed grenade flashes, etc.
 	var/list/worn_overlays = worn_overlays(standing, isinhands, file2use)
 	if(worn_overlays?.len)
+		if(!isinhands && default_layer && ishuman(loc))
```
Comment (jlsnow301) [inline code/modules/mob/living/carbon/human/human_update_icons.dm:714]: any way to extract this to break up the indentation a bit?

**Review by jlsnow301 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/human.dm [hunk 5]
```diff
@@ -759,6 +759,26 @@
...
+			HUMAN_HEIGHT_TALLEST
+		)
+		if(!(var_value in heights))
+			return
+
+		. = set_mob_height(var_value)
+
+	if(!isnull(.))
+		datum_flags |= DF_VAR_EDITED
+		return
```
Comment (Iamgoofball) [inline code/modules/mob/living/carbon/human/human.dm:778]: this is going to always return meaning ..() never calls and vv breaks for every other var

**Comment by ValuedEmployee** [comment]
> I can already tell that this can cause a cavalcade of issues with sprites. You can already tell the issues with the kneepads on the shortest one. Funny, but a pretty poor practical thing. image

In-game it makes moth wings look like they were clipped at the ends.
Dwarf1
Dwarf2
Dwarf3
And for that reason, I'm out.

**Comment by Mothblocks** [comment]
that looks fine enough

**Pull request merged**